### PR TITLE
depr(python): Rename `Expr.is_not()` to `Expr.not_()`

### LINKED ISSUE #10807: Rename `Expr.is_not()` to `Expr.not()` or `Expr.negate()`
### Problem description

The `Expr` object many methods that start with `is_`, e.g., `is_null()`, `is_in()`, and `is_first()`. You could say that they answer a question (with `True`s and `False`s).

The `Expr.is_not()` method, however, doesn't really answer a question, but performs an action&mdash;it negates a boolean expression.

For this reason I believe it would more consistent if the method `is_not()` would be renamed to `not()`. Another suggestion, if you'd prefer a verb, is to rename it to `negate()` (cf. PySpark[1] and the R package `purrr` [2]). 

Curious to hear what others think of this. 

_Side note: Python has the `is not` operator [3] (i.e., `operator.is_not(a, b)`) which checks for identity between two objects (and therefore does answer a question). I'm not saying that this causes confusion; just wanted to point it out._

### References

1. https://dist.apache.org/repos/dist/dev/spark/v3.5.0-rc1-docs/_site/api/python/reference/pyspark.sql/api/pyspark.sql.functions.negate.html#pyspark.sql.functions.negate
2. https://purrr.tidyverse.org/reference/negate.html
3.  https://docs.python.org/3/library/operator.html#operator.is_not

### PR DESCRIPTION
**Pull request by jeroenjanssens**
Closes #10807

As discussed in #10807.

### CODE REVIEW

**Comment by orlp** [comment]
Could you rebase on the latest `main`? Other than that looks good to me.

File: py-polars/docs/source/reference/expressions/boolean.rst [hunk 1]
```diff
@@ -16,8 +16,8 @@ Boolean
     Expr.is_infinite
     Expr.is_last
     Expr.is_nan
-    Expr.is_not
```
Comment (stinodego) [inline py-polars/docs/source/reference/expressions/boolean.rst:19]: We have to keep the documentation of the deprecated function.

File: py-polars/polars/expr/expr.py [hunk 2]
```diff
@@ -959,7 +959,7 @@ def pipe(
         '''
         return function(self, *args, **kwargs)
 
-    def is_not(self) -> Self:
+    def not_(self) -> Self:
```
Comment (stinodego) [inline py-polars/polars/expr/expr.py:962]: You need to add this new method in addition to the existing one, and deprecate the old method with `@deprectate_renamed_function`. Also add a `.. deprecated::` admonition to the old method.

You can check `Expr.apply` for an example.

File: crates/polars-plan/src/dsl/function_expr/boolean.rs [hunk 3]
```diff
@@ -81,27 +81,27 @@ impl From<BooleanFunction> for SpecialEq<Arc<dyn SeriesUdf>> {
     fn from(func: BooleanFunction) -> Self {
         use BooleanFunction::*;
         match func {
-            Any { ignore_nulls } => map!(any, ignore_nulls),
             All { ignore_nulls } => map!(all, ignore_nulls),
-            IsNot => map!(is_not),
-            IsNull => map!(is_null),
-            IsNotNull => map!(is_not_null),
+            AllHorizontal => wrap!(all_horizontal),
```
Comment (stinodego) [inline crates/polars-plan/src/dsl/function_expr/boolean.rs:85]: Please refrain from including random refactors like these in your PRs. While you might prefer alphabetical order, the functions were already ordered by function (e.g. IsFinite next to IsInfinite). Now we have to debate the merits of this change, while I really just want to review the deprecation...

**Review by stinodego — CHANGES_REQUESTED** [review]
I noticed `PySeries` has a `_not` method, but `Series` only has `__invert__`.

Maybe we should add `not_` to `Series` as well? And maybe we should add `__invert__` to `Expr`?

File: py-polars/docs/source/reference/expressions/boolean.rst (hunk 1 above)
Comment (jeroenjanssens) [reply py-polars/docs/source/reference/expressions/boolean.rst:19]: Thanks, will do.

File: crates/polars-plan/src/dsl/function_expr/boolean.rs (hunk 3 above)
Comment (jeroenjanssens) [reply crates/polars-plan/src/dsl/function_expr/boolean.rs:85]: Sorry about that. I'll revert those changes.

**Comment by jeroenjanssens** [comment]
> Thanks for the PR!

Thank you for all the feedback @stinodego.

> Maybe we should add `not_` to Series as well?

Sounds good to me. Where should I add this?

**Comment by stinodego** [comment]
> > Maybe we should add `not_` to Series as well?
> 
> Sounds good to me. Where should I add this?

I'd say rename the method on `PySeries` from `_not` to `not_`.

And then add a method on `Series`. You can add it anywhere it makes sense, there's no strict rules about that. Maybe before `is_null`.

And finish it off with a doc entry!

**Review by stinodego — APPROVED** [review]
I did a few things to finish this off:

* Add a doc entry for `Series.not_`
* Make `Series.__invert__` use `Series.not_`. Since that dispatches to the expression implementation, we can delete the Rust binding `PySeries.not_`.
* Add a test to make sure `is_not` still works and is properly deprecated.

Should be good to go, thanks for the initiative and work @jeroenjanssens !

**Comment by jeroenjanssens** [comment]
Thanks for your help, @stinodego! I'm happy I've been able to make another contribution (though small) to this amazing project.

**Comment by eitsupi** [comment]
Hi, I am uncomfortable with the use of `not_` instead of `not`, even on the Rust side, since the underscore suffix is a Python convention according to <https://github.com/pola-rs/polars/issues/10807#issuecomment-1703737552>'s explanation.
(Also, this is clearly a breaking change on the Rust side but was not so marked as that.)

Would you consider changing Rust's `not_` to `not`?

**Comment by stinodego** [comment]
If `not` is not reserved in Rust then that should be the name indeed.

**Pull request merged**